### ISSUE TITLE
Support styling as a hyperlink

### ISSUE THREAD

**Issue by mo8it** [ratatui-org/ratatui#1028]
See https://github.com/crossterm-rs/crossterm/issues/602 and https://gist.github.com/egmontkob/eb114294efbcd5adb1944c9f3cb5feda

**Comment by joshka** [comment]
I want this too. I suspect OSC 8 will just work if https://github.com/ratatui-org/ratatui/issues/902 is fixed.
A super high level on that issue is that the calculation of how many cells to skip needs to take into account that ANSI escape codes might mean that we write many symbols to the same cell.

A full solution to this would be to add a hyperlink widget that has properties for the text and hyperlink so app code can just do:
```rust
Hyperlink::new("Google", "https://google.com).render(area, buf);
```

Alternatively, I think as phrased you might be looking to add the hyperlink as another part of the `Style` struct, which would allow it to be added to any span/line/text without having to break the flow. Thoughts?

**Comment by mo8it** [comment]
Thanks for your quick reply 🤗

A widget would be too limiting. I think that this should be part of a `Paragraph` since such a hyperlink can be part of a longer text and might be broken/wrapped into multiple lines (so not only in one `Line`).

**Comment by joshka** [comment]
To be clear, a Hyperlink Widget as proposed above would accept anything that supports `Into<Text>`. I'd like eventually to move the wrapping into `Text` / `Line` / `Span` instead of wrapping being implemented in `Paragraph` only.

**Comment by joshka** [comment]
![Demo](https://vhs.charm.sh/vhs-2c5cvJzN7KTKR7U3Jbk71J.gif)

```rust
/// A hyperlink widget that renders a hyperlink in the terminal using [OSC 8].
///
/// [OSC 8]: https://gist.github.com/egmontkob/eb114294efbcd5adb1944c9f3cb5feda
struct Hyperlink<'content> {
    text: Text<'content>,
    url: String,
}

impl<'content> Hyperlink<'content> {
    fn new(text: impl Into<Text<'content>>, url: impl Into<String>) -> Self {
        Self {
            text: text.into(),
            url: url.into(),
        }
    }
}

impl WidgetRef for Hyperlink<'_> {
    fn render_ref(&self, area: Rect, buffer: &mut Buffer) {
        self.text.render_ref(area, buffer);
        let width = (area.width as usize).min(self.text.width());

        // this is a hacky workaround for https://github.com/ratatui-org/ratatui/issues/902, a bug
        // in the terminal code that incorrectly calculates the width of ANSI escape sequences. It
        // works by rendering the hyperlink as a series of 2-character chunks, which is the
        // calculated width of the hyperlink text.
        for (i, two_chars) in self
            .text
            .to_string()
            .chars()
            .chunks(2)
            .into_iter()
            .enumerate()
        {
            let text = two_chars.collect::<String>();
            let hyperlink = format!("\x1B]8;;{}\x07{}\x1B]8;;\x07", self.url, text);
            buffer
                .get_mut(area.x + i as u16 * 2, area.y)
                .set_symbol(hyperlink.as_str());
        }
    }
}
```

**Comment by joshka** [comment]
> A widget would be too limiting. I think that this should be part of a `Paragraph` since such a hyperlink can be part of a longer text and might be broken/wrapped into multiple lines (so not only in one `Line`).

Would you mind expanding a bit more on this point. I'm guessing with your rustlings work you have a bit more of an understanding of how Ratatui works now and might have thought more about how this idea might fit into things? Do you have any design ideas here?

**Comment by mo8it** [comment]
No, I didn't work long enough with Ratatui to be able to work on design ideas. I just wanted to mention that a link could be wrapped and therefore exist on two lines.

Example in a too small terminal pane:
![image](https://github.com/ratatui-org/ratatui/assets/76752051/03e2d194-f628-4773-92f7-1725e5b786a7)

**Comment by joshka** [comment]
If hyperlink was a part of Style, that would be represented as:
```rust
Line {
  spans: [
    Span::raw("A solution file can be found at "),
    Span::styled("solutions/quizzes/quiz2.rs", Style::new().hyperlink("solutions/quizzes/quiz2.rs")),
  ],
  ..Line::default()
}
```
(Lines are wrapped when put in a paragraph, OSC8 continues to the next line automatically AFAIK)

**Comment by felipesere** [comment]
How would I use this Hyperlink struct to create a `Cell` in a table?

**Comment by benjajaja** [comment]
Splitting up the OSC does not cut it, even now that we have `ForceWidth(1)` and we can really make one OSC8 per cell, because that just results in each individual letter being a link. Clicking on a letter opens a URL, but terminals usually put some special underline decorator when you hover the complete text. So when it's split, it underlines only one cell.

Alactritty:
<img width="116" height="32" alt="Image" src="https://github.com/user-attachments/assets/a4810a04-6028-4b5c-870b-8e1f8864568e" />

Kitty does group it, no clue why, but then there is a different bug where an artifact is drawn after the last link, and other OSC8 sequences elsewhere are also highlighted. WezTerm does something similar.
<img width="116" height="35" alt="Image" src="https://github.com/user-attachments/assets/5161c971-419e-4b04-8df8-14ace63ebc67" />

**Comment by cammeresi** [comment]
```rust
Line {
  spans: [
    Span::raw("A solution file can be found at "),
    Span::styled("solutions/quizzes/quiz2.rs", Style::new().hyperlink("solutions/quizzes/quiz2.rs")),
  ],
  ..Line::default()
}
```

I will try to add hyperlink support after crossterm makes a new release containing the newly accepted [hyperlink support](https://github.com/crossterm-rs/crossterm/pull/1038).

It would be helpful to have a preferred design nailed down.

What is proposed by @joshka looks very reasonable, but adding a link/string to Style will make it non-Copy, a breaking change.  Is that acceptable?

**Comment by joshka** [comment]
>Splitting up the OSC does not cut it

This was a hack workaround that worked briefly because of how ratatui handled an edge case. It wasn't really ever supposed to be the real solution to this.

**Comment by benjajaja** [comment]
Gotcha. Not sure how to solve this without making diffing/merging aware of spans though.

**Comment by VAWVAW** [comment]
> What is proposed by [@joshka](https://github.com/joshka) looks very reasonable, but adding a link/string to Style will make it non-Copy, a breaking change. Is that acceptable?

An alternative would be to add something like `ExtendedStyle` that is not `Copy` and implements `From<Style>`.

I'm not sure if changing the users of `Style` to accept `Into<ExtendedStyle>` would be breaking but we could also add function like `extended_styled()`.

**Comment by joshka** [comment]
> An alternative would be to add something like ExtendedStyle that is not Copy and implements From<Style>.

I'm wary of adding a bunch of hierarchical things - lest we end up with ExtendedSuperHyperStyle ;) (intentional slippery slope argument here, but perhaps game it out a bit more if you can).

**Comment by cammeresi** [comment]
I did a little exploratory work, and I don't think anyone's going to like removing `Copy` from `Style`.  I don't mind doing things other people don't like, and it offended even me.  As a sample of what happens:

- `fg`, `bg`, `add_modifier` et al. can no longer be `const`
- `Style` is passed by value in many, many places, which will require cloning
- widgets assume Style is `Copy`

It is significantly less intrusive (though still breaking) to hang the link off of `Span`, which already has to allocate.  That only takes about 400 lines of diff.

**Comment by cammeresi** [comment]
In case anyone cares, I have a provisional implementation in a fork at [cammeresi/ratatui](https://github.com/cammeresi/ratatui).

I have been successfully using this code in a private project, but I cannot post these patches until Crossterm makes a new release.

**Comment by natl-set** [comment]
is it working now?

**Comment by cammeresi** [comment]
> is it working now?

No, Crossterm is moving toward a release but hasn't done it yet.
